### DISCUSSION TITLE
Spark font rendering broken after AIR upgrade

### DISCUSSION THREAD

**Discussion by systemed** [airsdk/Adobe-Runtime-Support#3288] [Q&A]
I've recently updated my AIR version from 33.1.1 to 51.0.1. I'm overlaying it on Flex SDK 4.16.1, and building on a Mac.

Since the upgrade, any fonts rendered as part of a Spark component now show with the "missing character" glyph:

![Screenshot 2024-06-05 at 15 41 02](https://github.com/airsdk/Adobe-Runtime-Support/assets/694425/3b37fed9-262c-4e5c-9d37-b21addbcfb91)

In this example, the correctly rendered fonts are an `mx:TabNavigator` and an `mx:HSlider`. Everything else is Spark. The same is true throughout the application - any Spark text appears like this.

I'm not using any embedded fonts, just system Arial. That said, I've attempted to use embedded fonts to fix this and it doesn't make any difference: I can embed a .ttf and the mx text displays fine, but Spark is still like this, even with embedAsCFF=true.

If I build with the current version, but run with adl from 33.1, it's fine.

Any pointers to where I should look to debug this?

---

This minimal test case reproduces it:

minimal-app.xml:

```xml
<application xmlns="http://ns.adobe.com/air/application/33.1"> 
	<id>net.systemed.potlatch-minimal</id> 
	<versionNumber>3.1</versionNumber> 
	<filename>Potlatch</filename> 
	<initialWindow> 
		<content>minimal.swf</content> 
		<visible>true</visible> 
		<width>1200</width> 
		<height>800</height> 
		<renderMode>direct</renderMode>
		<requestedDisplayResolution>high</requestedDisplayResolution>
	</initialWindow> 
</application>
```

minimal.mxml:

```xml
<?xml version="1.0" encoding="utf-8"?>
<mx:Application
	xmlns:fx="http://ns.adobe.com/mxml/2009"
	xmlns:mx="library://ns.adobe.com/flex/mx"
	xmlns:s="library://ns.adobe.com/flex/spark"
	layout="vertical"
	horizontalScrollPolicy="off"
	verticalScrollPolicy="off"
	horizontalAlign="center" 
	addedToStage="startInit()" >

	<s:HGroup width="100%" verticalAlign="middle">
		<s:Label text="Simplify accuracy: "/>
		<mx:HSlider id="simplifyslider"
			minimum="1" maximum="5"
			dataTipPlacement="top" tickColor="black" labels="[High,Low]"
			snapInterval="1" tickInterval="1"
			allowTrackClick="true" liveDragging="true" />
	</s:HGroup>

	<fx:Script><![CDATA[
		private function startInit():void {
			trace("Inited app");
		}
	]]></fx:Script>

</mx:Application>
```

`amxmlc minimal.mxml -omit-trace-statements=false && adl minimal-app.xml`:

![Screenshot 2024-06-05 at 20 39 03](https://github.com/airsdk/Adobe-Runtime-Support/assets/694425/ce5efe8d-dc7f-4af3-b58a-a4701726dac3)

It's sufficiently basic that I'm sure it must be something to do with my build environment rather than the code, but I'm at a loss to see what.

**Comment by systemed** [comment]
Confirmed that this also happens with a completely fresh install of the AIR SDK (51.0.1) on an M2 Mac mini running macOS 14.1.1. 

Interestingly the Apache Flex installer itself seems to have the same issue:

![Screenshot 2024-06-06 at 07 26 48](https://github.com/airsdk/Adobe-Runtime-Support/assets/694425/998b0bd7-6bc7-4696-b8e5-215fa2784b58)

**Reply by pol2095** [reply]
Same issue on macOS, it work correctly using AIR SDK 50.2.5.1.

```
<?xml version="1.0" encoding="utf-8"?>
<s:Application xmlns:fx="http://ns.adobe.com/mxml/2009"    
    xmlns:s="library://ns.adobe.com/flex/spark">

	<s:TextArea width="300" text="Air Sdk"/>

</s:Application>
```

It's a TLF bug, I can reproduce it using [fl.text.TLFTextField](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/fl/text/TLFTextField.html) :

```
package
{
	import flash.display.Sprite;
	import fl.text.TLFTextField;
	
	public class Main extends Sprite
	{
		public function Main()
		{
			var textField:TLFTextField = new TLFTextField();
			textField.border = true;
			textField.text = "Air Sdk";
			this.addChild(textField);
		}
	}
}
```

**Reply by ajwfrost** [reply]
Seems like the macOS font loading code is different from the Windows code, in the handling of 'missing' tables. So we just need to make it a little more robust and cope with the fact that some fonts don't contain the tables that we added support for recently (for coloured emojis..)

Fix should be out in the next release, next week...

thanks

**Reply by systemed** [reply]
That's great - thank you. I can confirm that SDK 51.0.1.3 fixes this when bundled into a standalone app. Looking forward to the 51.0.1.3 shared runtime too.
